### PR TITLE
"Titanium Yellow" - Initial Mono.Native Support.

### PR DESCRIPTION
**Pull request by baulig**
`Mono.Native` is the new name for both `System.Native` and `System.Security.Cryptography.Apple`.
It is built as a stand-alone shared library and not bundled with the runtime because we
may need to build two different versions of it.

Starting with macOS 10.12+ and iOS 10+, Apple introduced a new Unified API for some of the
crypto primitives that we're using as part of System.Security.Cryptography.Apple.

On Desktop, we can check at runtime whether the OS version is recent enough and switch
implementation accordingly.  We build a single `libmono_native` shared library.

However, on Mobile we cannot have any undefined symbols as this would break Bitcode.

During the mobile build, we are called with `CC` containing an explicit minium version flag,
which is eiter `-mmacosx-version-min=`, `-mios-simulator-version-min=` or `-miphoneos-version-min=`
depending on platform.

We build two versions of the shared library:
- `libmono_native_compat` is built with whichever minimum version is passed to us via `CC`.
- `libmono_native_unifed` is built with the minimum version set to macOS 10.12+ / iOS 10+.

For testing purpuses, there is a function called `mono_native_get_platform_type ()`
(see mono/native/mono-native-platform.c), which returns a `MonoNativePlatformType` enum value.
There is also `Mono.MonoNativePlatform.GetPlatformType ()` (see mcs/class/corlib/Mono/MonoNativePlatform.cs).

This can be called by automated tests both to ensure that the library has been correctly installed and also
to verify that it's the correct version of it.

* `configure.ac`: add new configure checks for Mono.Native.

* `mono/native/`: new directory - this replaces the `System.Native` code that was previously in `mono/metadata`.

* `mcs/class/corlib/Mono`: New internal `MonoNativePlatformType` and `MonoNativePlatform` classes.

### CODE REVIEW

**Comment by vargaz** [comment]
Imho the libs should be named libmono-native-... instead of libmono_native, to conform to existing convention.

**Comment by baulig** [comment]
@vargaz Good point - and as it turns out, I actually did name them like that - so it's just wrong in the commit message and comments.

**Comment by baulig** [comment]
@luhenry I think the best way of addressing this question is assuming that we landed this PR before `System.Native` even existed, then asking how the "new" thing (`System.Native`) should fit it.  This logic isn't so far-fetched after all considering that from `xamarin-macios` point of view, `System.Native` actually _does not exist yet_.

The requirements for the "new" library will be as follows:

1. A mechanism of conditionally compiling C source files based on platform.
2. Bundling a shared library with a Xamarin.iOS / Xamarin.Mac app.
3. Bundling a shared library with a Xamarin.Android app.
4. Compiling C source files which come from an external source (corefx).

Optionally, these additional traits might be desirable:

6. Being able to conditionally bundle a shared library with a Xamarin.* app based on the existence of certain P/Invokes.
7. Being able to conditionally compile C source files based on target operating system versions.

Of these, 7. might become required when corefx adds code that requires newer version of macOS.  A potential iOS port of the `System.Native` code might also require different iOS versions.

If you look at all these requirements, then you'll see that the current approach already addresses 1., 2., 4. and 7., there's nothing in it that would conflict with implementing 3 - and 6. could be done at a later point once we've added more code to the native library and size becomes an issue.

We shouldn't look at this from the context of how do we bring this particular namespace / library (`System.Security.Cryptography.Apple` or `System.Native`), but rather address the problem as how do we deploy platform-specific C code from corefx to our products.

**Comment by baulig** [comment]
@monojenkins build

**Comment by luhenry** [comment]
This change will require modification in all products that depends on this existing `System.Native`, and I would expect you to do the work to modify the products accordingly. Also, that shows that this change is useless for Android for example since they don't have this problem, and it only seems to apply to iOS.

So we have 2 choices in front of us right now:
 1. keep `System.Native` as-is (you can rename it for all I care, but that will need to be reflected in the products, and that will not be left to whoever is integrating), and add a `System.Security.Cryptography.Apple` which will only be useful on iOS/macOS. That will require work only in XI/XM
 2. merge `System.Native` with `System.Security.Cryptography.Apple`. That will require work in XI/XM but also XA.

Since you will **have to** do the work in the products, I'll leave the choice to you 😉

**Comment by baulig** [comment]
Sure, I've already planned on doing all the work in the products when I started to work on this - and since XI/XM will be most complicated, I also volunteered to do the 2018-10 integration.  Don't worry about Android, I'll take care of all the System.Native work in there as well.

**Comment by baulig** [comment]
@monojenkins commit csproj

**Comment by baulig** [comment]
@monojenkins build failed

**Comment by baulig** [comment]
@monojenkins commit csproj

**Comment by baulig** [comment]
@monojenkins build Linux ARMv7

**Comment by baulig** [comment]
@monojenkins build Linux WebAssembly

**Comment by baulig** [comment]
@monojenkins build failed

**Comment by baulig** [comment]
@monojenkins build Linux ARMv5

File: mcs/class/System/Test/Mono/NativePlatformTest.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+		}
+
+		[Test]
+		public void Test ()
+		{
+			var type = mono_native_get_platform_type ();
+			Console.Error.WriteLine ($"NATIVE PLATFORM TYPE: {type}");
+			Assert.That (type, Is.GreaterThan (1), "platform type");
+
+			var usingCompat = (type & 16384) != 0;
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mcs/class/System/Test/Mono/NativePlatformTest.cs]: `MONO_NATIVE_PLATFORM_TYPE_COMPAT`?

File: configure.ac [hunk 2]
```diff
@@ -5483,6 +5483,190 @@ if test x"$GCC" = xyes; then
...
+#
+# However, on Mobile we cannot have any undefined symbols as this would break Bitcode.
+#
+# During the mobile build, we are called with `CC` containing an explicit minium version flag,
+# which is eiter `-mmacosx-version-min=`, `-mios-simulator-version-min=` or `-miphoneos-version-min=`
+# depending on platform.
+#
+# We build two versions of the shared library:
+# - `libmono_native_compat` is built with whichever minimum version is passed to us via `CC`.
+# - `libmono_native_unifed` is built with the minimum version set to macOS 10.12+ / iOS 10+.
```
[the remark this answers is not in the document]
Comment (baulig) [reply configure.ac]: We support those three configurations, but we don't actually build all three of them at the same time.  Depending on platform and minimum OS version selected, either `libmono-native` (desktop) _or_ the other two (mobile) are built.

Whenever we build without bitcode support, version checks at runtime can be performed.

For mobile builds with bitcode, we have the strict requirement of not having any undefined symbols - so any specific library cannot target any APIs that are not in whichever minimum OS version it's built against.

The requirement for "compat" is essentially "whichever is the oldest OS version that we still support".

For "unified", we can be more flexible and increase its minimum version requirement if needed (provided that a suitable "compat" implementation is available).

For your example, we have two options - either adding another "libmono-native-unified11" or making "unified" require iOS 11+.  We could essentially treat "unified" to target whichever OS we consider to be "modern" or "latest".

File: mcs/class/System/Test/Mono/NativePlatformTest.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (baulig) [reply mcs/class/System/Test/Mono/NativePlatformTest.cs]: The enum is internal in `System.dll`, so we can't access it in the tests.  I added a comment, though.

**Comment by baulig** [comment]
@monojenkins commit csproj

**Comment by baulig** [comment]
@monojenkins build failed

**Comment by baulig** [comment]
@monojenkins backport 2018-10

File: mcs/build/Makefile [hunk 3]
```diff
@@ -15,6 +15,14 @@ common/Consts.cs: common/Consts.cs.in $(wildcard config.make)
 	test -n '$(MONO_CORLIB_VERSION)'
 	sed -e 's,@''MONO_VERSION@,$(MONO_VERSION),' -e 's,@''MONO_CORLIB_VERSION@,$(MONO_CORLIB_VERSION),' $< > $@
 
+common/MonoNativeConfig.cs: common/MonoNativeConfig.cs.in $(wildcard config.make)
```
[the remark this answers is not in the document]
Comment (baulig) [reply mcs/build/Makefile]: @marek-safar @kg This is only used by one test.  I added a comment about that and disabled the test on all Windows platforms.

[the remark this answers is not in the document]
Comment (kg) [reply mcs/build/Makefile]: That might work, I'm not really sure. Please check the visual studio lane immediately after merging so we know whether it needs a fix.

[the remark this answers is not in the document]
Comment (baulig) [reply mcs/build/Makefile]: Sure, if anything breaks (which I don't expect), then I'll fix right away.

**Comment by baulig** [comment]
@monojenkins build OS X x64 iOS SDK

**Pull request merged**